### PR TITLE
feat: add MonadAsk & MonadReader instances

### PR DESCRIPTION
**Pull request by bentongxyz**
**Description of the change**

add MonadAsk & MonadReader instances

---

**Checklist:**

- [x] Added the change to the changelog's "Unreleased" section with a link to this PR and your username
- [ ] Linked any existing issues or proposals that this pull request should close
- [ ] Updated or added relevant documentation in the README and/or documentation directory
- [ ] Added a test for the contribution (if applicable)

### CODE REVIEW

File: src/Parsing.purs [hunk 1]
```diff
@@ -33,6 +33,7 @@ import Control.Alt (class Alt)
 import Control.Apply (lift2)
 import Control.Lazy (class Lazy)
 import Control.Monad.Error.Class (class MonadError, class MonadThrow, catchError, throwError)
+import Control.Monad.Reader (class MonadAsk, class MonadReader, ask, local)
```
Comment (jamesdbrock) [inline src/Parsing.purs:36]:
```suggestion
import Control.Monad.Reader.Class (class MonadAsk, class MonadReader, ask, local)
```

**Review by jamesdbrock — APPROVED** [review]
Thanks for the PR!

This looks correct to me, and I believe those instances can be very useful for `ParserT`. I'll leave the PR open for a few days to see if anyone else has an opinion.

**Review by garyb — APPROVED** [review]

File: src/Parsing.purs [hunk 2]
```diff
@@ -278,6 +279,15 @@ instance MonadRec (ParserT s m) where
 instance (MonadState t m) => MonadState t (ParserT s m) where
   state k = lift (state k)
 
+instance (MonadAsk r m) => MonadAsk r (ParserT s m) where
+  ask = lift ask
+
+instance (MonadRec m, MonadReader r m) => MonadReader r (ParserT s m) where
+  local f (ParserT k) = ParserT
+    ( mkFn5 \state1 more lift throw done ->
+        runFn5 k state1 more (lift <<< (local f)) throw done
```
Comment (natefaubion) [inline src/Parsing.purs:288]:
```suggestion
        runFn5 k state1 more (lift <<< local f) throw done
```
Unnecessary parens.

File: src/Parsing.purs [hunk 3]
```diff
@@ -278,6 +279,15 @@ instance MonadRec (ParserT s m) where
 instance (MonadState t m) => MonadState t (ParserT s m) where
   state k = lift (state k)
 
+instance (MonadAsk r m) => MonadAsk r (ParserT s m) where
+  ask = lift ask
+
+instance (MonadRec m, MonadReader r m) => MonadReader r (ParserT s m) where
```
Comment (natefaubion) [inline src/Parsing.purs:285]: Is the MonadRec constraint necessary?

Comment (bentongxyz) [reply src/Parsing.purs]: It is not, I have edit, thanks!

**Comment by jamesdbrock** [comment]
Thanks @bentongxyz !

**Comment by bentongxyz** [comment]
@jamesdbrock a joy to code on Purescript and this parsing library thanks to you all amazing people behind them. I am just happy to contribute something useful back!

**Pull request merged**